docs/knowledgebase/runtime/benchmarking.md

```diff
@@ -0,0 +1,547 @@
+
+    [features]
+    # -- snip --
+    runtime-benchmarks = ["frame-benchmarking"]
+    ```
+
+2. Start writing our benchmarking code. The code can be placed either in a `mod` module inside the
+   `src/lib.rs`, or be put in a separate file and include the module back in `src/lib.rs`. We will
+   use the first approach in the following to create the basic structure.
+
+    ```rust
+    #[cfg(feature = "runtime-benchmarks")]
+    mod benchmarking {
+      use super::*;
+      use frame_benchmarking::{benchmarks, account, impl_benchmark_test_suite};
+      use frame_system::RawOrigin;
+
+      benchmarks!{
+        // Individual benchmark cases here
+      }
+
+      impl_benchmark_test_suite!(Pallet, crate::tests::new_test_ext(),
+        crate::tests::Test);
+    }
+    ```
+
+3. Each benchmark case has three sections, setup section, execution section, and optionally a
+   verification section at the end.
+
+    ```rust
+    benchmarks!{
+      extrinsic_name {
+        // setup initial state
+      }: {
+        // running the benchmark
+      }
+      verify {
+        // Using assert statements to verify the result, or check certain on-chain event has been emitted.
+      }
+    }
+    ```
+
+    An example in
+    [`example/src/lib.rs`](https://github.com/paritytech/substrate/blob/master/frame/example/src/lib.rs):
+
+    ```rust
+    // This will measure the execution time of `set_dummy` for b in [1..1000] range.
+    set_dummy_benchmark {
+      let b in 1 .. 1000;
+    }: {
+      set_dummy (RawOrigin::Root, b.into());
+    }
+    ```
+
+    The name of the benchmark case is `set_dummy_benchmark`. Here `b` is the input that passed into
+    the extrinsic `set_dummy`. `b` will be varied from 1 to 1,000 and run repeatedly for each value
+    of `b` according to the arguments we passed when running the benchmark command. We will get back
+    to this in the following.
+
+    The function `set_dummy` is called in the execution section, and we do not verify the result at
+    the end.
+
+    You will also see code like this:
+
+    ```rust
+    accumulate_dummy {
+      let b in 1 .. 1000;
+      let caller = account("caller", 0, 0);
+    }: _ (RawOrigin::Signed(caller), b.into())
+    ```
+
+    This is when the execution section is just a one-line code that call the extrinsic with the same
+    name as the benchmarking case name. Our `benchmark!` macro is smart enough to know that!
+
+4.
+    ```rust
+    benchmarks! {
+      // -- snip
+      impl_benchmark_test_suite!(Pallet, crate::tests::new_test_ext(),
+        crate::tests::Test);
+    }
+    ```
+
+    At the last line in the `benchmarks!` macro, `impl_benchmark_test_suite!` takes three input: the
+    module name, the function that generates a test genesis storage `new_text_ext()`, and a test
+    runtime we built when writing our test cases, and expand them into testing code.
+
+    Ultimately benchmarking code is expanded as testing code, so it takes the code you write in
```

Comment:
I find this to be a bit confusing. The whole benchmarking does indeed _not_ expand to testing code, whereas the opposite, it should be run with `--release`. The testing portion is executed in a debug release.

As noted above, I think you can rework this section maybe, just to point out that: 
1. hey, you need a mock runtime for testing your benchmarks as well
2. hey, you might be able to re-use the one from your pallet testing
3. [additional links etc]

Reply:
I just finish reading the [benchmark! doc](https://substrate.dev/rustdocs/v3.0.0/frame_benchmarking/macro.benchmarks.html) and want to clarify what `impl_benchmark_test_suite!` does. 

- So running `substrate benchmark` actually doesn't run the code generated by `impl_benchmark_test_suite!`?
- It generates code to be run in `cargo test`, and run with the high and low values of  complexity parameters and see if it returns an `Ok(())`?

Reply:
>  So running substrate benchmark actually doesn't run the code generated by impl_benchmark_test_suite!?

No, that only runs the main benchmark code. The tests are run if you go into your pallet and run `cargo test --all-features`, or if you have done: 

```
#[cfg(any(test, feature = "runtime-benchmakr))]
mod benchmarking;
``` 

Then any `cargo test` will run the benchmark code as well. 

Some pallets do this, some don't afaik. I personally like this. The only downside is that if a benchmark test takes a lot of time, it will pollute the main test suite. But this is a rare case.

Reply:
> It generates code to be run in cargo test, and run with the high and low values of complexity parameters and see if it returns an Ok(())?

Correct. I think it generates just 3 steps per benchmark, with repeat=1, where the 3 steps are [low, mid, high] of the range.